library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/data/ExtraBlockData.java

```diff
@@ -0,0 +1,137 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Represents a {@linkplain BlockDataKey key} to value collection used to store extra data for {@link Block}s.
+ */
+@SuppressWarnings("ClassCanBeRecord")
+public final class ExtraBlockData {
+	private final Map<BlockDataKey<?>, Object> values;
+
+	private ExtraBlockData(Map<BlockDataKey<?>, Object> values) {
+		this.values = values;
+	}
+
+	/**
+	 * Constructs a new builder.
+	 *
+	 * @return new builder
+	 */
+	public static Builder builder() {
+		return new Builder();
+	}
+
+	/**
+	 * Gets or creates the {@code ExtraBlockData} collection that is tied to the specified block.
+	 *
+	 * @param block block
+	 * @return extra data collection
+	 */
+	public static ExtraBlockData getOrCreate(Block block) {
+		return QuiltBlockInternals.computeExtraData(block);
+	}
+
+	/**
+	 * Checks if this collection contains the specified key.
+	 *
+	 * @param key key to check
+	 * @return {@code true} if key has a value in this collection, {@code false} otherwise.
+	 */
+	public boolean contains(BlockDataKey<?> key) {
+		return values.containsKey(key);
+	}
+
+	/**
+	 * Gets a key's value.
+	 *
+	 * @param key key
+	 * @param <T> value type
+	 * @return value of key, or empty if value is missing.
+	 */
+	public <T> Optional<T> get(BlockDataKey<T> key) {
+		Object raw = values.get(key);
+		if (raw == null)
+			return Optional.empty();
+		if (!key.type().isInstance(raw))
+			throw new IllegalStateException("Value exists in collection, but type is incompatible with key type! Possible key collision?");
+		return Optional.of(key.type().cast(raw));
+	}
+
+	/**
+	 * A builder to construct a {@code ExtraBlockData} collection.
+	 */
+	public static final class Builder {
+		private final Map<BlockDataKey<?>, Object> values;
+
+		private Builder() {
+			values = new HashMap<>();
+		}
+
+		/**
+		 * Adds a key to value pair to the collection.
+		 *
+		 * @param key key
+		 * @param value value of key
+		 * @param <T> value type
+		 * @return this builder
+		 */
+		public <T> Builder put(BlockDataKey<T> key, T value) {
+			values.put(key, value);
+			return this;
+		}
+
+		/**
+		 * Builds the collection.
+		 *
+		 * @return new collection
+		 */
+		public ExtraBlockData build() {
+			return new ExtraBlockData(new HashMap<>(values));
+		}
+	}
+
+	/**
+	 * Invoked to compute an {@code ExtraBlockData} collection for a specific block.
+	 */
+	public interface OnBuild {
+		ArrayEvent<OnBuild> EVENT = ArrayEvent.create(OnBuild.class, callbacks -> (block, settings, builder) -> {
+			if (block instanceof OnBuild callback)
+				callback.append(block, settings, builder);
+			for (OnBuild callback : callbacks)
+				callback.append(block, settings, builder);
+		});
+
+		/**
+		 * Appends key to value pairs to the specified builder.
+		 *
+		 * @param block block
+		 * @param settings block settings
+		 * @param builder collection builder
+		 */
+		@ParameterInvokingEvent
```

Comment:
The `ParameterInvokingEvent` annotation is used incorrectly here. You're supposed to put it on a field that contains an event or a method which *returns* an event.

> ```
>  * <h2>How to use this annotation</h2>
>  *
>  * An event which may invoke a parameter should first place the annotation on the field of the event instance or the
>  * method which returns an event instance.
> ```

Reply:
Actually leo was correct here. Maybe I need to clarify it goes on the field/method which returns the event instance

Reply:
Never mind I was wrong